src/main/kotlin/com/openlattice/data/DataGraphService.kt

```diff
@@ -391,45 +401,61 @@ open class DataGraphService(
 
     override fun integrateAssociations(
             associations: Set<Association>,
-            authorizedPropertiesByEntitySet: Map<UUID, Map<UUID, PropertyType>>
+            authorizedPropertiesByEntitySetId: Map<UUID, Map<UUID, PropertyType>>
     ): Map<UUID, Map<String, UUID>> {
         val associationsByEntitySet = associations.groupBy { it.key.entitySetId }
         val entityKeys = HashSet<EntityKey>(3 * associations.size)
         val entityKeyIds = HashMap<EntityKey, UUID>(3 * associations.size)
 
+        //Create graph structure and check entity types
+        val associationEntitySetIds = mutableMapOf<UUID, MutableMap<UUID, MutableSet<UUID>>>() // edge-src-dst
+        val edges = associations
+                .asSequence()
+                .map { association ->
+                    val srcId = entityKeyIds[association.src]
+                    val dstId = entityKeyIds[association.dst]
+                    val edgeId = entityKeyIds[association.key]
+
+                    val srcEsId = association.src.entitySetId
+                    val dstEsId = association.dst.entitySetId
+                    val edgeEsId = association.key.entitySetId
+
+                    val src = EntityDataKey(srcEsId, srcId)
+                    val dst = EntityDataKey(dstEsId, dstId)
+                    val edge = EntityDataKey(edgeEsId, edgeId)
+
+                    if(associationEntitySetIds
+                                    .putIfAbsent(edgeEsId, mutableMapOf(srcEsId to mutableSetOf(dstEsId))) != null) {
+                        if(associationEntitySetIds.getValue(edgeEsId)
+                                        .putIfAbsent(srcEsId, mutableSetOf(dstEsId)) != null) {
+                            associationEntitySetIds.getValue(edgeEsId).getValue(srcEsId).add(dstEsId)
+                        }
+                    }
+
+                    DataEdgeKey(src, dst, edge)
+                }
+                .toSet()
+        checkAssociationEntityTypes(associationEntitySetIds)
```

Comment:
Similar to above you could make a call getAssociationEntityTypes and pass in that map to streamline look ups.